src/mlpack/methods/ann/layer/highway_impl.hpp

```diff
@@ -0,0 +1,246 @@
+    {
+      height = boost::apply_visitor(outputHeightVisitor, network.front());
+    }
+  }
+
+  for (size_t i = 1; i < network.size(); ++i)
+  {
+    if (!reset)
+    {
+      // Set the input width.
+      boost::apply_visitor(SetInputWidthVisitor(width), network[i]);
+
+      // Set the input height.
+      boost::apply_visitor(SetInputHeightVisitor(height), network[i]);
+    }
+
+    boost::apply_visitor(ForwardVisitor(std::move(boost::apply_visitor(
+        outputParameterVisitor, network[i - 1])), std::move(
+        boost::apply_visitor(outputParameterVisitor, network[i]))),
+        network[i]);
+
+    if (!reset)
+    {
+      // Get the output width.
+      if (boost::apply_visitor(outputWidthVisitor, network[i]) != 0)
+      {
+        width = boost::apply_visitor(outputWidthVisitor, network[i]);
+      }
+
+      // Get the output height.
+      if (boost::apply_visitor(outputHeightVisitor, network[i]) != 0)
+      {
+        height = boost::apply_visitor(outputHeightVisitor, network[i]);
+      }
+    }
+  }
+  if (!reset)
+  {
+    reset = true;
+  }
+
+  output = boost::apply_visitor(outputParameterVisitor, network.back());
+
+  if (arma::size(output) != arma::size(input)){
+    Log::Fatal << "The sizes of the output and input matrices of the Highway"
+        << " network should be equal. Please examine the network layers.";
+  }
+
+  transformGate = transformWeight * input;
+  transformGate.each_col() += transformBias;
+  transformGateActivation = 1.0 /(1 + arma::exp(-transformGate));
+  inputParameter = input;
+  networkOutput = output;
+  output = (output % transformGateActivation) +
+      (input % (1 - transformGateActivation));
+}
+
+template<typename InputDataType, typename OutputDataType,
+         typename... CustomLayers>
+template<typename eT>
+void Highway<InputDataType, OutputDataType, CustomLayers...>::Backward(
+    const arma::Mat<eT>&& input,
+    arma::Mat<eT>&& gy,
+    arma::Mat<eT>&& g)
+{
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, network.back())),
+      std::move(gy % transformGateActivation),
+      std::move(boost::apply_visitor(deltaVisitor, network.back()))),
+      network.back());
+
+  for (size_t i = 2; i < network.size() + 1; ++i)
+  {
+    boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+        outputParameterVisitor, network[network.size() - i])), std::move(
+        boost::apply_visitor(deltaVisitor, network[network.size() - i + 1])),
+        std::move(boost::apply_visitor(deltaVisitor,
+        network[network.size() - i]))), network[network.size() - i]);
+  }
+
+  g = boost::apply_visitor(deltaVisitor, network.front());
+
+  transformGateError = gy % (networkOutput - inputParameter) %
+      transformGateActivation % (1.0 - transformGateActivation);
+  g += transformWeight.t() * transformGateError;
+  g += gy % (1 - transformGateActivation);
```

Comment:
Hmm, can you explain what are you trying to achieve in the above code?

Reply:
That's the backward step for the highway layer. Since the output is `y = x % (1 - T(x)) + f(x) % T(x)` where T(x) is the transform gate activation. 
`dy/dx = (1 - T(x)) + x%d(1 - T(x))/dx + T(x)%df(x)/dx + f(x)%dT(x)/dx`
`dy/dx = (1 - T(x)) + T(x)df(x)/dx + (f(x)- x)%dT(x)/dx`
The first term is added on line 191, the last term on line 190 and the second term is calculated by backpropagation through the layers.

Reply:
Hmm, I'm pretty convinced of this approach, but I'd like to have @zoq's opinion on this as well.

Reply:
@zoq Any update on this?